### PR TITLE
[staging] pythonPackages.virtualenv: 20.2.1 -> 20.3.1

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/pull/109930 was reverted.

~This pull request is problematic cause it disabled virtualenv on Python 2.7, which is no longer supported by pytest-freezegun. I do not understand how virtualenv manages to run the tests anyway: https://github.com/pypa/virtualenv/blob/20.4.0/setup.cfg#L97.~
Solutions include
* packaging a separate version of pytest-freezegun for Python 2.7 or
* not running virtualenv's tests on Python 2.7.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pytest-freezegun/default.nix [hunk 1]
```diff
@@ -1,25 +1,34 @@
...
 
 buildPythonPackage rec {
   pname = "pytest-freezegun";
   version = "0.4.2";
 
-  src = fetchPypi {
-    inherit pname version;
-    extension = "zip";
-    sha256 = "19c82d5633751bf3ec92caa481fb5cffaac1787bd485f0df6436fd6242176949";
+  disabled = !isPy3k;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pytest-freezegun/default.nix:14]:
```suggestion
  disabled = isPy27;
```

File: pkgs/development/python-modules/virtualenv/default.nix [hunk 2]
```diff
@@ -47,10 +54,34 @@ buildPythonPackage rec {
...
+    cython
+    fish
+    flaky
+    pytest-freezegun
+    pytest-mock
+    pytest-timeout
+    pytestCheckHook
+  ];
+
+  preCheck = "export HOME=$(mktemp -d)";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/virtualenv/default.nix:67]:
```suggestion
  preCheck = ''
    export HOME=$(mktemp -d)
  '';
```

File: pkgs/development/python-modules/virtualenv/default.nix [hunk 3]
```diff
@@ -47,10 +54,34 @@ buildPythonPackage rec {
...
+
+  # Ignore tests which requires network access
+  pytestFlagsArray = [
+    "--ignore tests/unit/create/test_creator.py"
+    "--ignore tests/unit/seed/embed/test_bootstrap_link_via_app_data.py"
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/virtualenv/default.nix:73]: Please try out disabledTestFiles. It is the same as disabledTests but for files.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/virtualenv/default.nix (hunk 3 above)
Comment (dotlambda) [reply pkgs/development/python-modules/virtualenv/default.nix:73]: It does not work and there is no mention of it in `pkgs/development/interpreters/python/hooks/pytest-check-hook.sh`.

**Comment by dotlambda** [comment]
Any suggestions regarding Python 2.7?

File: pkgs/development/python-modules/virtualenv/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/virtualenv/default.nix:75]: It is only available on staging https://github.com/NixOS/nixpkgs/blob/staging/pkgs/development/interpreters/python/hooks/pytest-check-hook.sh#L5

**Comment by jonringer** [comment]
> Any suggestions regarding Python 2.7?

it's been EOL for over a year.

However, if I would "pin" the python2 version to 20.2.1 if it is necessary for the few remaining packages dependent on it.

**Comment by dotlambda** [comment]
I was able to avoid disabling `pytest-freezegun` on Python 2.7 by updating `python2Packages.freezegun`.

File: pkgs/development/python-modules/virtualenv/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/virtualenv/default.nix:75]: @dotlambda you forgot this.

Comment (dotlambda) [reply pkgs/development/python-modules/virtualenv/default.nix:75]: It doesn't work. See https://github.com/NixOS/nixpkgs/pull/108529#issuecomment-766361674.

**Comment by SuperSandro2000** [comment]
@jonringer can you please take another look at this?

**Comment by cole-h** [comment]
Adding `fish` as a `checkInput` to `python3Packages.virtualenv` means that everything that touches `fish` becomes a mass-rebuild for everything that depends on `python3Packages.virtualenv`. I've opened https://github.com/NixOS/nixpkgs/pull/114809 to remove this dependency.

**Pull request merged**